text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+  --> src/main.rs
+   |
+   |     for_each_city(do_something);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `FnMut` is not general enough
+   |
+   = note: `for<'a> fn(&'a str) -> impl Future<Output = ()> {do_something}` must implement `FnMut<(&str,)>`
+   = note: ...but it actually implements `FnMut<(&'0 str,)>`, for some specific lifetime `'0`
+```
+
+This happens because the type for the `Fut` generic parameter is chosen by the caller, in `main`, and it cannot reference the higher-ranked lifetime `for<'c>` in the `FnMut` trait bound, but the anonymous future produced by calling `do_something` does capture a generic lifetime parameter, and *must* capture it in order to use the `&str` argument.
+
+### Closures cannot return futures that borrow from their captures
+
+When users wants to *call* a function that takes an async callback argument, they often reach for `|| async {}` (a closure that returns an anonymous future) because closures can capture state from the local environment and are syntactically lightweight.
+
+However, users are quickly met with the limitation that they cannot use any of the closure's captures by reference in the async block.  E.g.:
+
+```rust
+async fn parent_country(city_name: &str) -> String { todo!() }
+
+async fn main() {
+    // Collect the country names of each city in our list.
+    let mut countries = vec![];
+    for_each_city(|city_name| async {
+        countries.push(parent_country(city_name).await);
+    })
+}
+```
+
+```
+error: captured variable cannot escape `FnMut` closure body
+  --> src/main.rs
+   |
+   |       let mut countries = vec![];
+   |           ------------- variable defined here
+   |       for_each_city(|city_name| async {
+   |  _____________________________-_^
+   | |                             |
+   | |                             inferred to be a `FnMut` closure
+   | |         countries.push(parent_country(city_name).await);
+   | |         --------- variable captured here
+   | |     }).await;
+   | |_____^ returns an `async` block that contains a reference to a captured variable, which then escapes the closure body
+   |
+   = note: `FnMut` closures only have access to their captured variables while they are executing...
+   = note: ...therefore, they cannot allow references to captured variables to escape
+```
+
+The future that is returned by the closure cannot reference any of the captures of the closure, which is a limitation that makes `|| async {}` quite unusable today without needing to, for example, clone data and declare the async block `move`.
+
+In order for this to work, the `FnMut` trait would need to be ["lending"](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern); however, there are [complications](https://hackmd.io/@compiler-errors/async-closures#Lending-closures-are-not-typically-FnOnce) with implementing general lending closures.
+
+# Guide Level Explanation
```

Comment:
Hi, thanks a lot for proposing this RFC first. It's a bit confusing for me to understand how `AsyncFnMut` works in reality.

For example, I built [`backon`](https://github.com/Xuanwo/backon): a async retry lib that allows user to retry an async function:

```rust
use anyhow::Result;
use backon::ExponentialBuilder;
use backon::Retryable;

async fn fetch() -> Result<String> {
    Ok(reqwest::get("https://www.rust-lang.org").await?.text().await?)
}

#[tokio::main]
async fn main() -> Result<()> {
    let content = fetch.retry(&ExponentialBuilder::default()).await?;
    println!("fetch succeeded: {}", content);

    Ok(())
}
```

The internal implemention involves `FnMut` and `Future` like:

https://github.com/Xuanwo/backon/blob/97f24cc51661fe3ded8578789d03529da403f664/src/retry.rs#L45-L55

```rust
pub trait Retryable<
    B: BackoffBuilder,
    T,
    E,
    Fut: Future<Output = Result<T, E>>,
    FutureFn: FnMut() -> Fut,
>
{
    /// Generate a new retry
    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, Fut, FutureFn>;
}
```

After this proposal implemented, I can write like this:

```rust
pub trait Retryable<
    B: BackoffBuilder,
    T,
    E,
    F: async FnMut() -> Result<T, E>,
>
{
    /// Generate a new retry
    fn retry(self, builder: &B) -> Retry<B::Backoff, T, E, F>;
}
```

Do I understand correctly? Would you like to provide an explanation of how `AsyncFnMut` is used? Thanks!

Reply:
> After this proposal implemented, I can write like this:

Yes.

Your mileage may vary if you need to, e.g., store the future returned by evaluating the future into a struct or something. See this [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async/topic/async.20iterator.20combinators.20using.20new.20async.20functionality/near/421775709) -- it's  not currently possible, but I'd like to add it eventually.

>  Would you like to provide an explanation of how AsyncFnMut is used?

I'm confused by this question -- isn't this the purpose of the RFC to explain how `AsyncFnMut` works? I recommend skipping over the "guide level explanation" and reading the detailed explanation section, then if you're confused about something specific, please point it out in the RFC.

Reply:
> I'm confused by this question -- isn't this the purpose of the RFC to explain how AsyncFnMut works?

Oh sorry. I mean adding an example of using `async FnMut()` as a generic type parameter in this section.

Reply:
Yeah sure